File: src/main/java/com/pestphp/pest/configuration/PestRunConfigurationHandler.java
package com.pestphp.pest.configuration;

import com.intellij.execution.ExecutionException;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.text.StringUtil;
import com.jetbrains.php.PhpBundle;
import com.jetbrains.php.config.commandLine.PhpCommandSettings;
import com.jetbrains.php.run.PhpExecutionUtil;
import com.jetbrains.php.testFramework.run.PhpTestRunConfigurationHandler;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class PestRunConfigurationHandler implements PhpTestRunConfigurationHandler {
    private static final Logger LOG = Logger.getInstance(PestRunConfigurationHandler.class);
    private static final PestRunConfigurationHandler instance = new PestRunConfigurationHandler();

    public static PestRunConfigurationHandler getInstance() {
        return instance;
    }

    @Override
    public @NotNull String getConfigFileOption() {
        return "--configuration";
    }

    @Override
    public void prepareCommand(@NotNull Project project, @NotNull PhpCommandSettings commandSettings, @NotNull String exe, @Nullable String version) throws ExecutionException {
        // String scriptFile = PhpExecutionUtil.loadHelperScriptAndGetText(project, "behat.php", commandSettings, PestRunConfigurationHandler.class);

        commandSettings.setScript(exe, false);
        commandSettings.addArgument("--teamcity");

        commandSettings.addEnv("IDE_PEST_EXE", exe);

        if (StringUtil.isNotEmpty(version)) {
            commandSettings.addEnv("IDE_PEST_VERSION", version);
        }
    }

    @Override
    public void runType(@NotNull Project project, @NotNull PhpCommandSettings phpCommandSettings, @NotNull String type, @NotNull String workingDirectory) throws ExecutionException {
        throw new ExecutionException("Can not run pest with type.");
    }

    @Override
    public void runDirectory(@NotNull Project project, @NotNull PhpCommandSettings phpCommandSettings, @NotNull String directory, @NotNull String workingDirectory) throws ExecutionException {
        phpCommandSettings.addPathArgument(directory);
    }

    @Override
    public void runFile(@NotNull Project project, @NotNull PhpCommandSettings phpCommandSettings, @NotNull String file, @NotNull String workingDirectory) throws ExecutionException {
        phpCommandSettings.addPathArgument(file);
    }

    @Override
    public void runMethod(@NotNull Project project, @NotNull PhpCommandSettings phpCommandSettings, @NotNull String filePath, @NotNull String methodName, @NotNull String workingDirectory) throws ExecutionException {
        throw new ExecutionException("Can not run pest with method.");
    }
}

File: src/main/kotlin/com/pestphp/pest/configuration/PestRunConfigurationHandler.kt
package com.pestphp.pest.configuration

import com.intellij.execution.ExecutionException
import com.intellij.openapi.project.Project
import com.jetbrains.php.config.commandLine.PhpCommandSettings
import com.jetbrains.php.testFramework.run.PhpTestRunConfigurationHandler
import com.pestphp.pest.PestBundle
import com.pestphp.pest.toPestTestRegex

class PestRunConfigurationHandler : PhpTestRunConfigurationHandler {
    var rootPath: String? = null

    companion object {
        @JvmField
        val instance = PestRunConfigurationHandler()
    }

    override fun getConfigFileOption(): String {
        return "--configuration"
    }

    override fun prepareCommand(project: Project, commandSettings: PhpCommandSettings, exe: String, version: String?) {
        commandSettings.setScript(exe, false)
        commandSettings.addArgument("--teamcity")
        commandSettings.addEnv("IDE_PEST_EXE", exe)

        if (!version.isNullOrEmpty()) {
            commandSettings.addEnv("IDE_PEST_VERSION", version)
        }
    }

    @Throws(ExecutionException::class)
    override fun runType(
        project: Project,
        phpCommandSettings: PhpCommandSettings,
        type: String,
        workingDirectory: String
    ) {
        throw ExecutionException(PestBundle.message("CANNOT_RUN_PEST_WITH_TYPE_MESSAGE"))
    }

    override fun runDirectory(
        project: Project,
        phpCommandSettings: PhpCommandSettings,
        directory: String,
        workingDirectory: String
    ) {
        if (directory.isEmpty()) {
            return
        }
        phpCommandSettings.addPathArgument(directory)
    }

    override fun runFile(
        project: Project,
        phpCommandSettings: PhpCommandSettings,
        file: String,
        workingDirectory: String
    ) {
        if (file.isEmpty()) {
            return
        }
        phpCommandSettings.addPathArgument(file)
    }

    override fun runMethod(
        project: Project,
        phpCommandSettings: PhpCommandSettings,
        file: String,
        methodName: String,
        workingDirectory: String
    ) {
        if (file.isEmpty()) {
            return
        }
        val pathMapper = phpCommandSettings.pathProcessor.createPathMapper(project)

        val rootPath = this.rootPath ?: workingDirectory

        phpCommandSettings.addPathArgument(file)
        phpCommandSettings.addArgument("--filter")
        phpCommandSettings.addArgument("/${methodName.toPestTestRegex(rootPath, file, pathMapper)}/")
    }
}
